mops/include/mops/utils.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef MOPS_UTILS_HPP
+#define MOPS_UTILS_HPP
+
+#include <cstddef>
+#include <cstdint>
+#include <vector>
+
+#include "mops/tensor.hpp"
+
+std::vector<std::vector<size_t>> get_write_list(mops::Tensor<int32_t, 1> write_coordinates);
+
+template <typename scalar_t> constexpr size_t get_simd_element_count();
+// Assume 256-bit vector registers. A conservative choice.
+// On HPE, 512 seems to provide an extra 50% boost on one core.
+template <> constexpr size_t get_simd_element_count<double>() {
+    return 256 / (sizeof(double) * 8); // 256 bits / 64 bits
+}
+template <> constexpr size_t get_simd_element_count<float>() {
+    return 256 / (sizeof(float) * 8); // 256 bits / 32 bits
+}
+
+template <typename scalar_t, size_t simd_element_count>
+void interleave_tensor(
+    mops::Tensor<scalar_t, 2> initial_data, scalar_t *interleft_data, scalar_t *remainder_data
+);
+
+template <typename scalar_t, size_t simd_element_count>
+void un_interleave_tensor(
+    mops::Tensor<scalar_t, 2> output_data, scalar_t *interleft_data, scalar_t *remainder_data
+);
+
+template <typename scalar_t, size_t simd_element_count>
+void interleave_tensor(
```

Comment:
why is the implementation separated from the declaration if both are in the same file?